core/arch/arm/mm/tee_mmu.c

```diff
@@ -680,15 +680,30 @@ static TEE_Result tee_mmu_user_va2pa_attr(const struct user_ta_ctx *utc,
 			if (pa) {
 				TEE_Result res;
 				paddr_t p;
+				size_t offset;
+
+				/*
+				 * Referenced mobj may not be not be page
```

Comment:
I can't parse this comment. what does "may not be not be page aligned" mean for instance?

I would have expected:
```
offset = utc->mmu->regions[n].offset + (vaddr_t)ua - utc->mmu->regions[n].va;
res = mobj_get_pa(utc->mmu->regions[n].mobj, offset, 0, pa);
```
to do the job.

Please explain which offset is added twice and why what I wrote above doesn't work.

Reply:
I agree, the comment is ugly.

The issue is that if the target buffer was not page alignment, `get_mobj_pa(region->mobj, region->offset)` returns the base phys addr of the buffer, which may already include the offset-in-page since buffer may start after the mapping based address.

`ua` obviously includes the offset-in-page value of the target addr to covert.
`region->va` is always map boundary aligned.

Hence, if we use `region->offset + ua - region->va` as offset in mobj_get_pa(), we twice the offset-in-page value.

Reply:
Thanks, finally I understand.
Wouldn't `CORE_MMU_USER_PARAM_{SIZE,MASK}` instead of `SMALL_PAGE_{SIZE,MASK}` be more appropriate?

Please try to improve the comment because this isn't very obvious.

Reply:
ok. thanks.
i will try to find the best readable code (and less commented code) way.
i am looking at the phys_to_virt and adding v2p / p2v in the embedded debug test.
the only place that currently test these are sdp+debug for v2p, as you mentioned, nothing tests p2v.